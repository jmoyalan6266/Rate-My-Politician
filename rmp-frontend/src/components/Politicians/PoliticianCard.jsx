import { Card, Button } from "react-bootstrap";
import { Link } from "react-router-dom";
import dummyHeadshot from "./dummyHeadshot.png";

const PoliticianCard = ({ politician, query }) => {
    const getHighlightedText = (text, highlight) => {
        // Split on highlight term and include term into parts, ignore case
        if (highlight) {
            const parts = text.toString().split(new RegExp(`(${highlight})`, 'gi'));
            return <span> { parts.map((part, i) => 
                <span key={i} style={part.toLowerCase() === highlight.toLowerCase() ? { backgroundColor: 'yellow' } : {} }>
                    { part }
                </span>)
            } </span>;
        } else {
            return text;
        }
    };

    return (
        <Card style={{ width: "20rem" }} className="h-100">
            <Card.Img variant="top" src={politician.photoURL ? politician.photoURL : dummyHeadshot} />
            <Card.Body>
                <center>
                    <Card.Title>{getHighlightedText(politician.name, query)}</Card.Title>
                    <Card.Text>
                        <strong>Party: </strong> {getHighlightedText(politician.party, query)}
                    </Card.Text>
                    <Card.Text>
                        <strong>State:</strong> {getHighlightedText(politician.state, query)}
                    </Card.Text>
                    <Card.Text>
                        <strong>Office:</strong> {getHighlightedText(politician.office, query)}
                    </Card.Text>
                    <Card.Text>
                        <strong>Office City:</strong> {getHighlightedText(politician.office_city, query)}
                    </Card.Text>
                    <Button as={Link} to={`/politicians/${politician.index}`}>
                        More Info
                    </Button>
                </center>
            </Card.Body>
        </Card>
    );
};

export default PoliticianCard;
