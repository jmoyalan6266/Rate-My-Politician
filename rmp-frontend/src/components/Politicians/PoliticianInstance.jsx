import {
    Container,
    Row,
    Col,
    Image,
    Button,
    Card,
    Jumbotron,
} from "react-bootstrap";
import { useParams } from "react-router";
import useSinglePolitician from "../../hooks/useSinglePolitician";
import useAxios from "axios-hooks";
import Loading from "../Loading";
import { Link } from "react-router-dom";
import dummyHeadshot from "./dummyHeadshot.png";
import { TwitterTimelineEmbed, TwitterFollowButton } from "react-twitter-embed";
import { IoIosArrowBack, IoIosArrowForward } from "react-icons/io";
import { useState } from "react";

const PoliticianInstance = () => {
    const { id } = useParams();
    const { politician, loading } = useSinglePolitician(id);
    const politicianState = politician.state;
    const [index, setIndex] = useState(0);

    const [{ data, isloading, error }, refetch] = useAxios(
        `${process.env.REACT_APP_API_URL}/politicians/${politician.name}/news`
    );

    const stateData = useAxios(
        `${process.env.REACT_APP_API_URL}/states/${politicianState}`
    );

    return (
        <Container className="my-4">
            {loading ? (
                <Loading />
            ) : (
                <div>
                    <center>
                        <h1 className="my-4">{politician.name}</h1>
                    </center>
                    <Row>
                        <Image
                            src={
                                politician.photoURL
                                    ? politician.photoURL
                                    : dummyHeadshot
                            }
                            style={{ maxHeight: "12rem" }}
                            className="mx-auto"
                            onError={(e) => {
                                e.target.onerror = null;
                                e.target.src = dummyHeadshot;
                            }}
                        />
                    </Row>
                    <Container className="mt-5">
                        <Jumbotron>
                            <Row>
                                <Card className="mx-auto my-2">
                                    <Card.Header>Party:</Card.Header>
                                    <Card.Body className="mx-auto">
                                        <Card.Text>{politician.party}</Card.Text>
                                    </Card.Body>
                                </Card>
                                <Card className="mx-auto my-2">
                                    <Card.Header>State:</Card.Header>
                                    <Card.Body className="mx-auto">
                                        <Card.Text>{politician.state}</Card.Text>
                                    </Card.Body>
                                </Card>
                                <Card className="mx-auto my-2">
                                    <Card.Header>Office:</Card.Header>
                                    <Card.Body className="mx-auto">
                                        <Card.Text>{politician.office}</Card.Text>
                                    </Card.Body>
                                </Card>
                                <Card className="mx-auto my-2">
                                    <Card.Header>Office City:</Card.Header>
                                    <Card.Body className="mx-auto">
                                        <Card.Text>
                                            {politician.office_city}
                                        </Card.Text>
                                    </Card.Body>
                                </Card>
                                <Card className="mx-auto my-2">
                                    <Card.Header>
                                        Number of News Articles:
                                    </Card.Header>
                                    <Card.Body className="mx-auto">
                                        <Card.Text>{data?.totalResults}</Card.Text>
                                    </Card.Body>
                                </Card>
                            </Row>
                            <br></br>
                            <Row>
                                <Card className="mx-auto my-2">
                                    <Card.Header>Phone:</Card.Header>
                                    <Card.Body className="mx-auto">
                                        <Card.Text>
                                            {politician.phone_number}
                                        </Card.Text>
                                    </Card.Body>
                                </Card>
                                <Card className="mx-auto my-2">
                                    <Card.Header>Website:</Card.Header>
                                    <Card.Body className="mx-auto">
                                        <Card.Text>
                                            <a href={politician.website}>
                                                {politician.website}
                                            </a>
                                        </Card.Text>
                                    </Card.Body>
                                </Card>
                                <Card className="mx-auto my-2">
                                    <Card.Header>Twitter</Card.Header>
                                    <Card.Body className="mx-auto">
                                        <Card.Text>
                                            <TwitterFollowButton
                                                screenName={politician.twitter}
                                                options={{
                                                    height: 100,
                                                    width: 100,
                                                }}
                                            />
                                        </Card.Text>
                                    </Card.Body>
                                </Card>
                                <Card className="mx-auto my-2">
                                    <Card.Header>Facebook</Card.Header>
                                    <Card.Body className="mx-auto">
                                        <Card.Text>{politician.facebook}</Card.Text>
                                    </Card.Body>
                                </Card>
                            </Row>
                        </Jumbotron>
                    </Container>
                    <br></br>
                    <Row>
                        <Col>
                            <TwitterTimelineEmbed
                                sourceType="profile"
                                screenName={politician.twitter}
                                options={{ height: 500, width: 800 }}
                            />
                        </Col>
                        <Col>
                            <br></br>
                            <center>
                                <h1>Learn More About This Politician's State</h1>
                            </center>
                            <br></br>
                            <Row className="my-4">
                                <Col>
                                    <Card
                                        className="ml-auto mr-auto"
                                        style={{ width: "16rem" }}
                                    >
                                        {stateData[0].loading ? (
                                            <Loading />
                                        ) : (
                                            <Card.Img
                                                src={
                                                    stateData[0].data
                                                        ?.state_flag_url
                                                }
                                            />
                                        )}
                                        <Card.Header
                                            as={Link}
                                            to={`/states/${politicianState}`}
                                        >
                                            {politicianState}
                                        </Card.Header>
                                    </Card>
                                </Col>
                            </Row>
                        </Col>
                    </Row>
                    <br></br>
                    <Row>
                        <center>
                            <h1>News Article Mentions</h1>
                            <Row>
                                <IoIosArrowBack
                                    style={{ marginTop: "15rem" }}
                                    onClick={() =>
                                        index > 0 ? setIndex(index - 1) : null
                                    }
                                    size="2.5em"
                                />
                                {data?.articles
                                    .slice(index, index + 4)
                                    .map((article, i) => (
                                        <Col style={{width: "10%"}} key={i}>
                                            <Card
                                                key={i}
                                                style={{ width: "100%" }}
                                                className="mx-auto m-4"
                                            >
                                                <Card.Header>
                                                    <Link
                                                        to={`/news/${article.source.name}`}
                                                    >
                                                        <strong>
                                                            {article.source.name}
                                                        </strong>
                                                    </Link>
                                                </Card.Header>

                                                <Card.Img
                                                    src={article.urlToImage}
                                                    variant="top"
                                                />
                                                <Card.Body>
                                                    <Card.Text className="p-3">
                                                        {article.title}
                                                    </Card.Text>
                                                    <Button
                                                        as="a"
                                                        href={article.url}
                                                        size="sm"
                                                    >
                                                        Article Link
                                                    </Button>
                                                </Card.Body>
                                            </Card>
                                        </Col>
                                    ))}
                                <IoIosArrowForward
                                    style={{ marginTop: "15rem" }}
                                    onClick={() =>
                                        index < 19 ? setIndex(index + 1) : null
                                    }
                                    size="2.5em"
                                />
                            </Row>
                        </center>
                    </Row>
                </div>
            )}
        </Container>
    );
};

export default PoliticianInstance;
