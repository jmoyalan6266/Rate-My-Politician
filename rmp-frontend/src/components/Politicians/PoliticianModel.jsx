import React, { Component, forwardRef } from "react";
import { useHistory } from "react-router-dom";
import MaterialTable from "material-table";
import {
    AddBox,
    ArrowDownward,
    Check,
    ChevronLeft,
    ChevronRight,
    Clear,
    DeleteOutline,
    Edit,
    FilterList,
    FirstPage,
    LastPage,
    Remove,
    SaveAlt,
    Search,
    ViewColumn,
} from "@material-ui/icons";
import usePoliticians from "../../hooks/usePoliticians";
import { Link } from "react-router-dom";
import dummyHeadshot from "./dummyHeadshot.png";
import republicanParty from "./republicanParty.png";
import democraticParty from "./democraticParty.png";
import {
    Button,
} from "react-bootstrap";

const tableIcons = {
    Add: forwardRef((props, ref) => <AddBox {...props} ref={ref} />),
    Check: forwardRef((props, ref) => <Check {...props} ref={ref} />),
    Clear: forwardRef((props, ref) => <Clear {...props} ref={ref} />),
    Delete: forwardRef((props, ref) => <DeleteOutline {...props} ref={ref} />),
    DetailPanel: forwardRef((props, ref) => (
        <ChevronRight {...props} ref={ref} />
    )),
    Edit: forwardRef((props, ref) => <Edit {...props} ref={ref} />),
    Export: forwardRef((props, ref) => <SaveAlt {...props} ref={ref} />),
    Filter: forwardRef((props, ref) => <FilterList {...props} ref={ref} />),
    FirstPage: forwardRef((props, ref) => <FirstPage {...props} ref={ref} />),
    LastPage: forwardRef((props, ref) => <LastPage {...props} ref={ref} />),
    NextPage: forwardRef((props, ref) => <ChevronRight {...props} ref={ref} />),
    PreviousPage: forwardRef((props, ref) => (
        <ChevronLeft {...props} ref={ref} />
    )),
    ResetSearch: forwardRef((props, ref) => <Clear {...props} ref={ref} />),
    Search: forwardRef((props, ref) => <Search {...props} ref={ref} />),
    SortArrow: forwardRef((props, ref) => (
        <ArrowDownward {...props} ref={ref} />
    )),
    ThirdStateCheck: forwardRef((props, ref) => (
        <Remove {...props} ref={ref} />
    )),
    ViewColumn: forwardRef((props, ref) => <ViewColumn {...props} ref={ref} />),
};
function Politicians() {
    const { allPoliticians, loading } = usePoliticians();
    let history = useHistory();
    const divStyle = {
        position: 'relative',
        display: 'inline',
    };
    return (
        <div
            style={{
                minHeight: "900px",
                margin: "auto",
                paddingTop: "20px",
                width: "80%",
            }}
            className="mb-4"
        >
            <center>
                <h1 className="mb-3 mt-1">Politicians</h1>
            </center>
            <MaterialTable
                icons={tableIcons}
                title="Politicians"
                data={allPoliticians}
                columns={[
                    {
                        title: "",

                        render: (politician) => (
                            <div style={divStyle}>
                                <img src={
                                    politician.party == 'Democratic Party'
                                        ? democraticParty
                                        : politician.party == 'Republican Party'
                                            ? republicanParty
                                            : null
                                }
                                    width="20"
                                    height="20"
                                    style={{position: 'absolute'}}
                                />

                                <img
                                    src={
                                        politician.photoURL
                                            ? politician.photoURL
                                            : dummyHeadshot
                                    }
                                    onError={(e) => {
                                        e.target.onerror = null
                                        e.target.src = dummyHeadshot
                                      }}
                                    class="img-fluid"
                                    style={{ 
                                        width: 20 + "vw"
                                        
                                    }}
                                    width="35%"
                                />
                            </div>
                        )
                    },
                    {
                        title: "Name",

                        field: "name",
                    },
                    {
                        title: "Party",
                        field: "party",
                    },
                    {
                        title: "State",
                        field: "state",
                    },
                    {
                        title: "Office",
                        field: "office",
                    },
                    {
                        title: "Office City",
                        field: "office_city",
                    },
                    {
                        title: "",
                        render: (politician) => (
                            <td>
                                <Button variant="outline-dark"
                                    as={Link}
                                    to={`/politicians/${politician.index}`}
                                    style={ politician.party == 'Democratic Party'
                                    ? { color: "white", background: "blue" }
                                    : politician.party == 'Republican Party'
                                        ? { color: "white", background: "red" }
                                        : { color: "black", background: "white" }}
                                >
                                    More Info
                                </Button>
                            </td>
                        ),
                    },
                ]}
                options={{
                    //Link to instance page
                    onRowClick: (rowData) => {
                        history.push("/politicians/" + rowData[0]);
                    },
                    sorting: true,
                    filtering: true,
                    showTitle: false,
                    rowStyle: (x) => {
                        if (x.tableData.id % 2) {
                            return { backgroundColor: "#f2f2f2" };
                        }
                    },
                    pageSize: 10,
                    pageSizeOptions: [10, 20, 50],
                    headerStyle: {
                        backgroundColor: "#6797cf",
                        fontSize: 16,
                    },
                }}
            />
        </div>
    );
}
export default Politicians;
