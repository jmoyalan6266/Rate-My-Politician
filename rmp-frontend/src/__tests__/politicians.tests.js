import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from "enzyme-adapter-react-16"
import Politicians from "../components/Politicians/PoliticianModel"

configure({ adapter: new Adapter() })

describe("Test Politician Model page views", () => {
	test("Politicians", () => {
		const politiciansTest = shallow(<Politicians />)
		expect(politiciansTest).toMatchSnapshot()
	})
})
