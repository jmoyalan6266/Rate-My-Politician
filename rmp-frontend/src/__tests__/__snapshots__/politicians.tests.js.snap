// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Politician Model page views Politicians 1`] = `
<div
  className="mb-4"
  style={
    Object {
      "margin": "auto",
      "minHeight": "900px",
      "paddingTop": "20px",
      "width": "80%",
    }
  }
>
  <center>
    <h1
      className="mb-3 mt-1"
    >
      Politicians
    </h1>
  </center>
  <WithStyles(Component)
    columns={
      Array [
        Object {
          "render": [Function],
          "title": "",
        },
        Object {
          "field": "name",
          "title": "Name",
        },
        Object {
          "field": "party",
          "title": "Party",
        },
        Object {
          "field": "state",
          "title": "State",
        },
        Object {
          "field": "office",
          "title": "Office",
        },
        Object {
          "field": "office_city",
          "title": "Office City",
        },
        Object {
          "render": [Function],
          "title": "",
        },
      ]
    }
    data={Array []}
    icons={
      Object {
        "Add": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "Check": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "Clear": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "Delete": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "DetailPanel": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "Edit": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "Export": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "Filter": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "FirstPage": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "LastPage": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "NextPage": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "PreviousPage": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "ResetSearch": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "Search": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "SortArrow": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "ThirdStateCheck": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
        "ViewColumn": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
      }
    }
    options={
      Object {
        "filtering": true,
        "headerStyle": Object {
          "backgroundColor": "#6797cf",
          "fontSize": 16,
        },
        "onRowClick": [Function],
        "pageSize": 10,
        "pageSizeOptions": Array [
          10,
          20,
          50,
        ],
        "rowStyle": [Function],
        "showTitle": false,
        "sorting": true,
      }
    }
    title="Politicians"
  />
</div>
`;
